plugins/cryptolib/cryptolib.cpp

```diff
@@ -527,6 +527,54 @@ CRYPTOLIB_API bool CRYPTOLIB_CALL clPKIVerifySignatureBuff(const char * pkalgori
     }
 }
 
+//-----------------------------------------------------------------
+//  Simple cache for loaded keys
+//-----------------------------------------------------------------
+class CKeyCache
+{
+private:
+    CriticalSection csKeyCache;//guards serial modifications to the cache map
+    typedef std::unordered_map<string, Owned<CLoadedKey>> KeyCache;
+    KeyCache keyCache;
+
+public:
+    CLoadedKey * getInstance(bool isPublic, const char * keyFS, const char * keyBuff, const char * passphrase)//
+    {
+        VStringBuffer searchKey("%s_%s_%s", isEmptyString(keyFS) ? "" : keyFS, isEmptyString(keyBuff) ? "" : keyBuff, isEmptyString(passphrase) ? "" : passphrase);
```

Comment:
Shouldn't we bail out and avoid entering the CritBlock if keyFS and KeyBuff are empty?

Reply:
Done